platform/security/idp/security-idp-server/src/main/java/org/codice/ddf/security/idp/server/IdpEndpoint.java

```diff
@@ -228,37 +233,61 @@ public void init() {
-            try {
-                MetadataConfigurationParser metadataConfigurationParser =
-                        new MetadataConfigurationParser(serviceProviderMetadata, ed -> {
-                            EntityInformation entityInfo = new EntityInformation.Builder(ed,
-                                    SUPPORTED_BINDINGS).build();
-                            if (entityInfo != null) {
-                                serviceProviders.put(ed.getEntityID(), entityInfo);
-                            }
-                        });
-
-                serviceProviders.putAll(metadataConfigurationParser.getEntryDescriptions()
-                        .entrySet()
-                        .stream()
-                        .map(e -> Maps.immutableEntry(e.getKey(),
-                                new EntityInformation.Builder(e.getValue(),
-                                        SUPPORTED_BINDINGS).build()))
-                        .filter(e -> nonNull(e.getValue()))
-                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
-
-            } catch (IOException e) {
-                LOGGER.warn("Unable to parse SP metadata configuration. Check the configuration for SP metadata.", e);
+    private Map<String, EntityInformation> getServiceProvidersMap() {
+        Map<String, EntityInformation> spMap = serviceProviders.get();
+        if (spMap == null) {
+            spMap = parseServiceProviderMetadata();
+
+            boolean updated = serviceProviders.compareAndSet(null, spMap);
+            if (!updated) {
+                LOGGER.debug("Safe but concurrent update to serviceProviders map; using processed value");
             }
         }
+
+        return Collections.unmodifiableMap(spMap);
+    }
+
+    private Map<String, EntityInformation> parseServiceProviderMetadata() {
+        if (spMetadata == null) {
+            return Collections.emptyMap();
+        }
+
+        Map<String, EntityInformation> spMap = new ConcurrentHashMap<>();
+        try {
+            MetadataConfigurationParser metadataConfigurationParser =
+                    new MetadataConfigurationParser(spMetadata, ed -> {
+                        EntityInformation entityInfo = new EntityInformation.Builder(ed,
+                                SUPPORTED_BINDINGS).build();
+                        if (entityInfo != null) {
+                            spMap.put(ed.getEntityID(), entityInfo);
+                        }
+                    });
+
+            spMap.putAll(metadataConfigurationParser.getEntryDescriptions()
+                    .entrySet()
+                    .stream()
+                    .map(e -> Maps.immutableEntry(e.getKey(),
+                            new EntityInformation.Builder(e.getValue(),
+                                    SUPPORTED_BINDINGS).build()))
+                    .filter(e -> nonNull(e.getValue()))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
```

Comment:
I don't see how that's possible. They keys making it to the `Collector` are the same keys that were in the `metadataConfigurationParser.getEntryDescriptions` map. They're distinct on the start of the stream.